### ISSUE TITLE
compile problem of reentrant in XC8

### ISSUE THREAD

**Issue by ZanoZ** [cocoOS/cocoOS#5]
Dear:
I have compile the cocoos in xc8, , the original version is ok, 
and it report an error, after I add the os_tick() into timer ISR.
the error said that there are reentrant issue, and blablabla.... when I "\\" the os_tick(), it will fine.
this is the error:"os_task.c:478: advisory: (1510) non-reentrant function "_os_task_clear_wait_queue" appears in multiple call graphs and has been duplicated by the compiler"

I know the different if reentrant function and usual function. But I am not sure it is right or not to add reentrant before every function in the code...
I don't know why you guys works well. maybe the compiler is different...
I use mplabx and xc8 v1.44
Thanks for any following..

BR
Zona.W

**Comment by ZanoZ** [comment]
I now have followed a few step into the functions, and found that the call-tree is crosee on task_clear_wait_queue()...
I changed the announce of this function to :   
reentrant void os_task_clear_wait_queue( uint8_t tid );
compile passed but it is said that the stack may overflow...
I think the RIGHT way is that don't make too much functions in the os_tick(), let the timer ISR returned and 
do the task switch in main line . I think it is a better/more safe.
or another way is to make a different call tree of os_tick for timer isr,...
[10min later]
I made a little steps to duplicate the os_tick, and found it is not a good choice to make a different call tree of os_tick, the functions do too much on the kernal , it is not safe to make changed both in main line  and in ISR function call line...
I am thinking ,..
ANY comments will be helpful. thanks for reading .

**Comment by ZanoZ** [comment]
After a few hours stepping in the code, and thinking ...:
1, add "reentrant" before the funtion  is not a right way, to solve the problem.(I means the call-tree crossing...
2, there should be a lock , witch should be locked when any OS functions started to make the task schedule or do anything on tasklist[];, and when ISR call os_tick() , the os_tick() should check the lock, if it is locked, skip the works on tick or tasklist...  this may cause more delay on the tasks, but safe.
 further, the os core can be more exquisite , the lock is placed in deeper functions to reduce the time of locking period.
Hope I am right. I still not familiar to this os and unable to make the change. 
Hope you can read it soon. Dear @CocoOs @ptrxt @ivankravets

**Comment by cocoOS** [comment]
Hello Zano,
I am not familiar with your compiler. Have never seen those errors you have run into. 
I only use gcc. Can't help you for the moment. I'm sorry.

**Comment by ZanoZ** [comment]
Hi cocoOS,
It ok, I can/have to  do a little work in code reading, and I have already got the success in compile.but still have some trouble in coding. 
And my idea is that the cocoOS can't rely on the compiler to make the code running well, It should make the code itself right and safe. Any comment are welcome.

for other following here, I share how to cover when you meet the trouble in "call-tree crossing" issue:
As files said, in timer ISR , the os_tick() is used to make the tick, but some compiler can't solve this problem("call-tree crossing") well, so my solution is :
1, before main(), add  Sem_t bin_OSTCK_sem;
2, create a new task , I named it:
                   void Task_OSTick(void){
		task_open();
		for(;;)
		{
		os_tick();
		 sem_wait(bin_OSTCK_sem);
		}
		task_close();
             }
3, set the Task_OSTick lowest prio in your all tasks.
4, in timer isr , use 	sem_ISR_signal(bin_OSTCK_sem); instead os_tick();
.end
Above is tested in MPLABX with compiler 1.4x.

hope cocoOS helps all~

**Comment by matteofranceschini** [comment]
Hi ZanoZ,
the **"os_task.c:478: advisory: (1510) non-reentrant function "_os_task_clear_wait_queue" appears in multiple call graphs and has been duplicated by the compiler"** problem is absolutely normal and it depends on the fact that the os_task_clear_wait_queue is a function that is called both from interrupt and "normal code": the compiler does not like that and therefore it creates TWO different functions to keep the calls independent. You don't have to worry about that.

**Comment by ZanoZ** [comment]
yes you are right, but duplicate the function may cause the function running on the same time.(when main function running the interrupt also running this function), and will cause the function have a wrong result.
Because of this risk, I think i should be very careful on this issue.

**Comment by matteofranceschini** [comment]
@ZanoZ what you're saying is actually incorrect!
The compiler is duplicating the function **exactly** because it's dangerous to have the **same** function being called from the interrupt routine **and** normal code.

Refer to the [XC8 user manual](http://ww1.microchip.com/downloads/en/DeviceDoc/MPLAB_XC8_C_Compiler_User_Guide_for_PIC.pdf), page 141, for detailed description of the problem. You can of course disable this feature (as the manual states) but you have to make sure that every time you call the function the interrupts are disabled **before**.

Hope this helps clarifying.

**Comment by ZanoZ** [comment]
well, what will happen in following code:
------
unsigned int temp;
main()
{
...
while (1)
   { temp++;
     delay_ms(1000);
     functionA();
   }
}
functionA()
{ .....
    temp=temp+1000;       //Line point A
.....
}
interrupt ISR();
{
    functionA();
}

well, when the mcu running in the "Line point A", it is working on temp=temp+1000, the ISR happens,
also the another functionA() works, 
after ISR, the temp value is already changed, and feedback to temp;
but if you view the detail of temp=temp+1000; it should be something like Reg1=temp, Reg2=Reg1+1000, temp=Reg2,, so it maybe have risk to lost the ISR's working resulte (after the ISR save back the temp+1000 value back).

ANYWAY, I must say that I have not tested/verify my idea, but it has a risk and I should not put the risk on compiler's work.
----er....my english is really pool, hope you can understand what I am typing,...

**Comment by matteofranceschini** [comment]
hi @ZanoZ , check at page 140 of the overmentioned XC8 compiler manual:

> It is assumed by the compiler that an interrupt can occur at any time. Functions
encoded to use the compiled stack are not reentrant (see Section 4.5.2.2.1 “Object
Size Limits”), so if such a function is called by an interrupt function and by main-line
code, this could lead to code failure.

> MPLAB XC8 compiler has a feature which will duplicate the generated code associated
with any function that uses the non-reentrant function model and which is called from
more than one call graph. There is one call graph associated with main-line code, and
one for each interrupt function, if defined. It is assumed by the compiler that an interrupt
may occur at any time. This allows reentrancy, but recursion is still not possible, even
if the function is duplicated.

As far as the manual explains, it's not strictly a race condition issue: the problem relies in the way the stack and the memory of the PIC is configured.

**Comment by ZanoZ** [comment]
Hi matteofranceschini, Thanks for your words.
I understand the manual . But i am not sure if the compiler can make this well.
Maybe I should make a test program to verify it.
